### DISCUSSION TITLE
undefined reference to `__device_dts_ord_XX'

### DISCUSSION THREAD

**Discussion by Dave-Poissant** [zephyrproject-rtos/zephyr#50628] [Q&A] [answered]
Hi zephyr developers, Hello!

I'm currently trying to reproduce the "blinky pwm" sample onto my stm32f411ce and I'm facing different problems that I would like to clear asap. I'm working with the platformio ide env with zephyr release 2.7.1.

At this moment, I have made a pwm device as follows in my overlay file:

```
/ {
    aliases {
            i2c1node = &i2c1;
            i2c3node = &i2c3;
            spi1node = &spi1;
            uartnode = &usart2;
            pwmled0 = &pwm_led_white;
    };

    pwmleds {
        compatible = "pwm-leds";
        pwm_led_white: pwm_led_white {
            label = "PWM_LED_WHITE";
            pwms = < &pwm1 1 256 PWM_POLARITY_NORMAL >;
        };
    };
};

&timers1 {
    status = "okay";
    pwm1: pwm {
        status = "okay";
        pinctrl-0 = <&tim1_ch1n_pb13>;
        pinctrl-names = "default";
    };
};
```

And the output in the zephyr.dts is:

```
aliases {
		led0 = &user_led;
		sw0 = &user_button;
		i2c1node = &i2c1;
		i2c3node = &i2c3;
		spi1node = &spi1;
		uartnode = &usart2;
		pwmled0 = &pwm_led_white;
	};

{...}

timers1: timers@40010000 {
	compatible = "st,stm32-timers";
	reg = < 0x40010000 0x400 >;
	clocks = < &rcc 0x3 0x1 >;
	interrupts = < 0x18 0x0 >, < 0x19 0x0 >, < 0x1a 0x0 >, < 0x1b 0x0 >;
	interrupt-names = "brk", "up", "trgcom", "cc";
	status = "okay";
	label = "TIMERS_1";
	pwm1: pwm {
		compatible = "st,stm32-pwm";
		status = "okay";
		st,prescaler = < 0x0 >;
		label = "PWM_1";
		#pwm-cells = < 0x3 >;
		pinctrl-0 = < &tim1_ch1n_pb13 >;
		pinctrl-names = "default";
		phandle = < 0x1c >;
	};
};

{...}

pwmleds {
		compatible = "pwm-leds";
		pwm_led_white: pwm_led_white {
			label = "PWM_LED_WHITE";
			pwms = < &pwm1 0x1 0x100 0x0 >;
		};
	};
```

Now, I'm able to get the device and its infos out of the ```DT_NODELABEL(pwmled0)``` maccros like so:

```
#define PWM_LED0_NODE	DT_NODELABEL(pwm_led_white)

#if DT_NODE_HAS_STATUS(PWM_LED0_NODE, okay)
#define PWM_CTLR	DT_PWMS_CTLR(PWM_LED0_NODE)
#define PWM_CHANNEL	DT_PWMS_CHANNEL(PWM_LED0_NODE)
#define PWM_FLAGS	DT_PWMS_FLAGS(PWM_LED0_NODE)
#else
#error "Unsupported board: pwmled0 devicetree alias is not defined"
#define PWM_CTLR	DT_INVALID_NODE
#define PWM_CHANNEL	0
#define PWM_FLAGS	0
#endif
```

but whenever I try to use these defines in my code I get a build error saying:
```undefined reference to `__device_dts_ord_26'```

Here is an example of how I want to use the PWM_LED0_NODE define:

```
void main(void)
{
    {...}
    const struct device *pwm_dev = DEVICE_DT_GET(PWM_LED0_NODE);
    if (!pwm_dev)
    {
    SWOPrintf("Cannot find PWM device!\n\r");
    return;
    }
    else
    {
    SWOPrintf("PWM device find\n\r");
    SWOPrintf("%s\n\r", pwm_dev->name);
    }
    
    if (!device_is_ready(pwm_dev)) {
    SWOPrintf("Error: PWM device %s is not ready\n\r", pwm_dev->name);
    return;
    }
}
```

I hope I was exhaustive enough, I'm still getting used to this RTOS and to devicetree so don't mind asking for more infos if there is something missing.

Thx a lot!

**Comment by Dave-Poissant** [comment]
Has an addition I'm putting my prj.conf in here:
```
# nothing here
CONFIG_LOG=y
CONFIG_LOG_PRINTK=y

CONFIG_DMA=y
CONFIG_UART_STM32=y
CONFIG_SERIAL=y
CONFIG_UART_ASYNC_API=y

# enable pinmux
CONFIG_PINMUX=y
CONFIG_PWM=y

# CONFIG_UART_INTERRUPT_DRIVEN=y

CONFIG_GPIO=y
```

**Comment by erwango** [comment]
`undefined reference to `__device_dts_ord_26'` is generally a missing Kconfig symbol. `CONFIG_LED_PWM=y` in your case, maybe ?

**Reply by Dave-Poissant** [reply]
I added CONFIG_LED_PWM=y but I obtain the same ```undefined reference to``` error.

**Reply by erwango** [reply]
Ok, looking the other way round: I can see you enabled `CONFIG_DMA=y` did you enabled a DMA node in device tree ?

**Reply by Dave-Poissant** [reply]
yep, i had to enabled them  in my overlay to use CONFIG_UART_ASYNC on my stm32:
```
&dma1 {
    status = "okay";
};

&dma2 {
    status = "okay";
};
```

**Reply by Dave-Poissant** [reply]
But its indeed weird, because if i reference the "pwm1" device instead of the "pwm_led_white" i get no errors. But when i set the pwm with this func:
```
pwm_pin_set_usec(pwm_dev, 1, 256, 200, PWM_POLARITY_NORMAL);
```
i get no output on my pin.

**Comment by Dave-Poissant** [comment, marked as answer]
Well I'm not sure if this was the problem, but I made some changes to my "pwmleds" property:
```
aliases {
            i2c1node = &i2c1;
            i2c3node = &i2c3;
            spi1node = &spi1;
            uartnode = &usart2;
            pwmled0 = &pwm_led_white;
    };

    pwmleds {
        compatible = "pwm-leds";
        pwm_led_white: external_pwm_led {
            pwms = < &pwm1 1 256 PWM_POLARITY_NORMAL >;
        };
    };
```

So now the device name "pwm_led_white" is different then the property's name "external_pwm_led". I seem to be able to extract the node with its alias. @erwango  correct me if I'm wrong but there is a binding named "external_pwm_leds" in "pwmleds" that I was not using before?

I still have no voltage outputted to my pin though, but this must be related to another problem through my initialization of the driver. I think my question is answered, I'll make another question for the absence of voltage at my pin.
